r/tests/testthat/test-duckdb.R

```diff
@@ -15,6 +15,19 @@
 # specific language governing permissions and limitations
 # under the License.
 
+test_that("meaningful error message when duckdb is not installed", {
+  new_lib <- tempfile()
+  dir.create(new_lib)
+  withr::with_libpaths(new_lib, {
+    ds <- InMemoryDataset$create(example_data)
+    expect_error(
+      to_duckdb(ds),
+      "Please install the `duckdb` package."
+    )
+  })
+  unlink(new_lib, recursive = TRUE)
```

Comment:
Nice use of `with_libpaths()` and nice that this is up at the top before duckdb might have already been loaded (we should probably also have a comment explaining that since it might be confusing to others who run across this code).

`.libPaths()` is a strange beast that has some pretty peculiar behavior. I suspect what's going on here is some combination of:
* that `new_lib` might not actually be created as a directory yet. (if you pass `.libPaths()` a string that's not a directory it silently ignores it)
* `.libPaths()` won't overwrite or add to libraries specified in `.Library.site` or `.Library` https://stat.ethz.ch/R-manual/R-devel/library/base/html/libPaths.html

You might want to try and print `.libPaths()` to confirm what you expect it to be. You also might want to use our helper `make_temp_dir()` which should ensure that a temp directory is actually created

Reply:
`new_lib` works correctly and gets inserted at the first position of `.libPaths()`.
``` r
new_lib <- tempfile()
dir.create(new_lib)
withr::with_libpaths(new_lib, {
  .libPaths()
})
#> [1] "/private/var/folders/wx/zjw1jzk11s76vchhc2xym7xw0000gn/T/RtmpHwjdcB/file1ed63169505e"
#> [2] "/Library/Frameworks/R.framework/Versions/4.1/Resources/library"
unlink(new_lib, recursive = TRUE)
```

<sup>Created on 2022-02-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

I think the issue is that the system library will always be at `.libPaths()[[2]]`, which means the code above should work provided you have _not_ installed packages in the system library.

Reply:
I think the question might be how to make sure there's only one `dir` on `.libPaths()` for the purpose of running that test. I believe {rent} might be doing just that.

https://github.com/rstudio/renv/blob/86edb40cf638c017f26c7f9be155bbbbaf01b1b0/R/libpaths.R#L12-L13

Reply:
> which means the code above should work provided you have not installed packages in the system library.

Yeah, exactly. You might try (temporarily) changing the values of `.Library.site` and `.Library` inside of this test and see if that will trigger it.

Alternatively, we have a number of tests that specifically don't install duckdb on them, maybe we could skip this test if duckdb *is* installed, and then run the test only if it's not installed (e.g. test-r-ubuntu-21.04).  https://github.com/apache/arrow/blob/50ed858635f421d370420b63f30b38cf42bb6103/ci/scripts/r_deps.sh#L33-L40

Reply:
An option could be to add 
```r
dir.create(Sys.getenv("R_LIBS_USER"), recursive=TRUE)
```

to our CI jobs before installing any dependencies. That way we make sure anything we install will not be in the system library. `R_LIBS_USER` should by default be at `.libPaths()[1]` unless it doesn't exists (when it is silently ignored)

Reply:
I think the most straightforward solution would be to skip the test when `duckdb` is installed.